File: apps/api/matematica_api/services/github_models.py
import os
from typing import Any, Dict, List, Optional, Tuple

import httpx
from urllib.parse import urlparse


class GithubModelsClient:
    def __init__(self) -> None:
        self.token = os.getenv("GITHUB_TOKEN")
        self.model = os.getenv("GITHUB_MODELS_MODEL")
        self.endpoint = os.getenv(
            "GITHUB_MODELS_ENDPOINT",
            "https://models.github.ai/inference/chat/completions",
        )

    def is_configured(self) -> bool:
        return bool(self.token and self.model)

    def chat(self, messages: List[Dict[str, Any]]) -> Tuple[Optional[str], Optional[str]]:
        if not self.is_configured():
            return None, "GitHub Models credentials are not set. Provide GITHUB_TOKEN and GITHUB_MODELS_MODEL to enable chat."

        parsed = urlparse(self.endpoint)
        is_azure = "azure.com" in (parsed.netloc or "")

        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
        }

        if is_azure:
            headers["api-key"] = self.token or ""
        else:
            headers.update(
                {
                    "Authorization": f"Bearer {self.token}",
                    "Accept": "application/vnd.github+json",
                    "X-GitHub-Api-Version": "2022-11-28",
                }
            )
        payload = {
            "model": self.model,
            "messages": messages,
            "temperature": 0.2,
        }
        try:
            response = httpx.post(self.endpoint, headers=headers, json=payload, timeout=30)
            if response.status_code >= 400:
                detail = response.text
                return None, f"GitHub Models request failed ({response.status_code}): {detail}"
            response.raise_for_status()
            data = response.json()
            choice = data.get("choices", [{}])[0]
            message = choice.get("message", {})
            return message.get("content", ""), None
        except Exception as exc:  # noqa: BLE001
            return None, f"GitHub Models request failed: {exc}"

File: lib/github-models.ts
export interface GitHubModelsUsage {
  promptTokens?: number;
  completionTokens?: number;
  totalTokens?: number;
}

export interface GitHubModelsChatResult {
  content: string;
  usage?: GitHubModelsUsage;
}

export async function githubModelsChat(
  messages: Array<{ role: string; content: string }>,
  token?: string,
  model?: string,
  endpoint?: string,
  options?: {
    temperature?: number;
    maxTokens?: number;
  }
): Promise<string> {
  const result = await githubModelsChatWithUsage(messages, token, model, endpoint, options);
  return result.content;
}

export async function githubModelsChatWithUsage(
  messages: Array<{ role: string; content: string }>,
  token?: string,
  model?: string,
  endpoint?: string,
  options?: {
    temperature?: number;
    maxTokens?: number;
  }
): Promise<GitHubModelsChatResult> {
  const apiToken = token || process.env.GITHUB_TOKEN;
  const apiModel = model || process.env.GITHUB_MODELS_MODEL;
  const apiEndpoint =
    endpoint ||
    process.env.GITHUB_MODELS_ENDPOINT ||
    "https://models.github.ai/inference/chat/completions";

  if (!apiToken || !apiModel) {
    throw new Error(
      "GitHub Models credentials not configured. Set GITHUB_TOKEN and GITHUB_MODELS_MODEL env vars."
    );
  }

  const isAzure = apiEndpoint.includes("azure.com");

  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    Accept: "application/json",
  };

  if (isAzure) {
    headers["api-key"] = apiToken;
  } else {
    headers["Authorization"] = `Bearer ${apiToken}`;
    headers["Accept"] = "application/vnd.github+json";
    headers["X-GitHub-Api-Version"] = "2022-11-28";
  }

  const payload = {
    model: apiModel,
    messages,
    ...(!apiModel.includes("gpt-5") && typeof options?.temperature === "number"
      ? { temperature: options.temperature }
      : {}),
    ...(typeof options?.maxTokens === "number"
      ? apiModel.includes("gpt-5")
        ? { max_completion_tokens: options.maxTokens }
        : { max_tokens: options.maxTokens }
      : {}),
  };

  try {
    const response = await fetch(apiEndpoint, {
      method: "POST",
      headers,
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      const text = await response.text();
      throw new Error(
        `GitHub Models request failed (${response.status}): ${text}`
      );
    }

    const data = await response.json();
    const choice = data.choices?.[0];
    const message = choice?.message;
    const content = message?.content || "";
    const usage = data?.usage
      ? {
          promptTokens: data.usage.prompt_tokens,
          completionTokens: data.usage.completion_tokens,
          totalTokens: data.usage.total_tokens,
        }
      : undefined;

    return { content, usage };
  } catch (error) {
    throw new Error(
      `GitHub Models request failed: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}
